File: VeritasStorageOptimizer/engine/veritas_cleaner.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Veritas Engine - Local Storage Cleaner & Optimizer (2026 Production Edition)
Architecture: Standard Library Only / UTF-8 No BOM / Stream-Safe
"""
import os
import sys
import hashlib
import argparse
from pathlib import Path
from datetime import datetime

# 預設保護名單與暫存檔檔名標籤
TEMP_EXTENSIONS = {".tmp", ".log", ".cache", ".bak", ".old", ".temp", ".swp", ".dmp"}
PROTECTED_DIRS = {".git", ".svn", ".venv", "node_modules", "site-packages", "$RECYCLE.BIN", "System Volume Information"}
DUP_MIN_BYTES = 1024  # 重複比對的最小檔案容量門檻


def is_skipped_dir(parent: str, name: str) -> bool:
    """保護名單、symlink,以及任何含 pyvenv.cfg 的 Python 虛擬環境一律跳過:
    各 venv 內含成千上萬相同的套件檔案,跨 venv 去重會毀掉第一個以外的環境。"""
    if name in PROTECTED_DIRS:
        return True
    dir_path = Path(parent) / name
    if dir_path.is_symlink():
        return True
    return (dir_path / "pyvenv.cfg").is_file()


def calculate_hash_stream(file_path: Path, chunk_size: int = 65536) -> str:
    """採用串流讀取計算 MD5,避免記憶體溢位"""
    hasher = hashlib.md5()
    try:
        with open(file_path, "rb") as f:
            while chunk := f.read(chunk_size):
                hasher.update(chunk)
        return hasher.hexdigest()
    except (PermissionError, OSError):
        return ""


def format_size(size_bytes: float) -> str:
    """轉換容量格式為可讀字串"""
    for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
        if size_bytes < 1024.0:
            return f"{size_bytes:.2f} {unit}"
        size_bytes /= 1024.0
    return f"{size_bytes:.2f} PB"


def clean_empty_directories(target_path: Path, dry_run: bool = True) -> list:
    """由下而上遞迴清理空資料夾"""
    removed_dirs = []
    for root, dirs, _ in os.walk(target_path, topdown=False):
        for d in dirs:
            dir_path = Path(root) / d
            if is_skipped_dir(root, d):
                continue
            try:
                if not any(dir_path.iterdir()):
                    removed_dirs.append(dir_path)
                    if not dry_run:
                        dir_path.rmdir()
            except (PermissionError, OSError):
                continue
    return removed_dirs


def execute_cleanup(target_dir: str, size_threshold_mb: float, dry_run: bool = True,
                    log_file: str = "cleanup_audit.log"):
    target_path = Path(target_dir).resolve()
    if not target_path.exists() or not target_path.is_dir():
        print(f"[Error] Target directory '{target_dir}' does not exist.")
        sys.exit(1)

    print(f"=== Starting Scan: {target_path} (Mode: {'DRY-RUN' if dry_run else 'EXECUTE'}) ===", flush=True)
    size_limit_bytes = int(size_threshold_mb * 1024 * 1024)
    size_groups = {}
    files_to_remove = []
    freed_bytes = 0

    # Pass 1: 遍歷檔案與處理暫存檔 / 超大檔案
    for root, dirs, files in os.walk(target_path):
        # 排除保護目錄 / symlink / Python 虛擬環境
        dirs[:] = [d for d in dirs if not is_skipped_dir(root, d)]

        for file in files:
            file_path = Path(root) / file

            try:
                file_size = file_path.stat().st_size
            except (PermissionError, OSError):
                continue

            # 策略 1: 刪除副檔名為暫存檔的檔案
            if file_path.suffix.lower() in TEMP_EXTENSIONS:
                files_to_remove.append((file_path, file_size, "Temp File"))
                freed_bytes += file_size
                continue

            # 策略 2: 標記超過閾值的大檔案 (非暫存檔)
            if file_size > size_limit_bytes:
                files_to_remove.append((file_path, file_size, f"Over Limit (>{size_threshold_mb}MB)"))
                freed_bytes += file_size
                continue

            # 策略 3: 為重複檔案比對收集「檔案大小相同的群組」
            # 小於 DUP_MIN_BYTES 的檔案不列入:刪除近乎零容量的「重複檔」
            # 釋放不了空間,卻常是 __init__.py / .gitkeep 等結構性檔案
            if file_size >= DUP_MIN_BYTES:
                size_groups.setdefault(file_size, []).append(file_path)

    # Pass 2: 針對容量相同的候選群組進行 Hash 比對
    for file_size, candidate_paths in size_groups.items():
        if len(candidate_paths) < 2:
            continue

        hashes = {}
        for path in candidate_paths:
            h = calculate_hash_stream(path)
            if not h:
                continue
            if h in hashes:
                files_to_remove.append((path, file_size, f"Duplicate of {hashes[h].name}"))
                freed_bytes += file_size
            else:
                hashes[h] = path

    # Pass 3: 執行刪除作業與日誌紀錄
    audit_logs = [
        "=== VERITAS CLEANUP AUDIT LOG ===",
        f"Timestamp: {datetime.now().isoformat()}",
        f"Target Directory: {target_path}",
        f"Execution Mode: {'DRY-RUN' if dry_run else 'LIVE_DELETE'}\n",
        "--- Target File Items ---"
    ]

    for path, size, reason in files_to_remove:
        line = f"[{reason}] {format_size(size)} -> {path}"
        audit_logs.append(line)
        print(line, flush=True)
        if not dry_run:
            try:
                path.unlink()
            except Exception as e:
                audit_logs.append(f"  └─ Failed to remove: {e}")

    removed_empty_dirs = clean_empty_directories(target_path, dry_run=dry_run)
    audit_logs.append("\n--- Removed Empty Directories ---")
    for d in removed_empty_dirs:
        audit_logs.append(f"[Empty Directory] -> {d}")
        print(f"[Empty Directory] -> {d}", flush=True)

    audit_logs.append("\nSummary:")
    audit_logs.append(f"Total Files Marked: {len(files_to_remove)}")
    audit_logs.append(f"Total Empty Dirs Marked: {len(removed_empty_dirs)}")
    audit_logs.append(f"Estimated Space Freed: {format_size(freed_bytes)}")

    # 寫入稽核日誌
    with open(log_file, "w", encoding="utf-8") as f:
        f.write("\n".join(audit_logs))

    print("Scan complete.")
    print(f"Items to purge: {len(files_to_remove)} files, {len(removed_empty_dirs)} empty folders.")
    print(f"Total Space to Free: {format_size(freed_bytes)}")
    print(f"Audit log saved to: {Path(log_file).resolve()}")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Veritas Standard Storage Optimizer 2026")
    parser.add_argument("--dir", type=str, required=True, help="Target folder path to clean")
    parser.add_argument("--max-mb", type=float, default=200.0, help="File size limit threshold in MB")
    parser.add_argument("--log", type=str, default="cleanup_audit.log", help="Audit log output path")
    parser.add_argument("--execute", action="store_true",
                        help="Perform live deletion (Defaults to Dry-Run if omitted)")

    args = parser.parse_args()
    if not args.max_mb > 0:
        parser.error("--max-mb must be a positive number")
    execute_cleanup(target_dir=args.dir, size_threshold_mb=args.max_mb,
                    dry_run=not args.execute, log_file=args.log)

File: VeritasStorageOptimizer/engine/veritas_cleaner.js
#!/usr/bin/env node
/**
 * Veritas Engine - Node.js High-Performance File Purger (2026 Edition)
 * Zero-Dependency / Asynchronous Stream Hashing / Non-Blocking
 */
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

const TEMP_EXTENSIONS = new Set(['.tmp', '.log', '.cache', '.bak', '.old', '.temp', '.swp', '.dmp']);
const PROTECTED_DIRS = new Set(['.git', '.svn', '.venv', 'node_modules', 'site-packages', '$RECYCLE.BIN', 'System Volume Information']);
// Files below this size never enter duplicate detection: freeing ~0 bytes,
// while near-empty twins (__init__.py, .gitkeep, stdout stubs) are structural.
const DUP_MIN_BYTES = 1024;

// Coding byproducts: caches and compiled leftovers that tooling regenerates.
const CODING_JUNK_DIRS = new Set(['__pycache__', '.pytest_cache', '.mypy_cache', '.ruff_cache', '.ipynb_checkpoints']);
const CODING_JUNK_EXTENSIONS = new Set(['.pyc', '.pyo']);

// User-placed marker: any directory containing this file is off-limits to
// every strategy, letting users pin "do not touch" zones inside a scan.
const PROTECT_MARKER = '.veritas_protect';

// Protected names, Python virtual environments (pyvenv.cfg marker), and
// user-protected zones: venvs share thousands of identical package files,
// so cross-venv dedup would gut every env after the first-scanned one.
function isSkippedDir(fullPath, name) {
    if (PROTECTED_DIRS.has(name)) return true;
    try {
        return fs.existsSync(path.join(fullPath, 'pyvenv.cfg'))
            || fs.existsSync(path.join(fullPath, PROTECT_MARKER));
    } catch (err) { return true; }
}

function formatSize(bytes) {
    const units = ['B', 'KB', 'MB', 'GB', 'TB'];
    let size = bytes;
    let i = 0;
    while (size >= 1024 && i < units.length - 1) {
        size /= 1024;
        i++;
    }
    return `${size.toFixed(2)} ${units[i]}`;
}

// 串流雜湊計算,防止大型檔案直接卡爆 V8 Buffer
function getFileHashAsync(filePath) {
    return new Promise((resolve) => {
        const hash = crypto.createHash('md5');
        const stream = fs.createReadStream(filePath);
        stream.on('data', chunk => hash.update(chunk));
        stream.on('end', () => resolve(hash.digest('hex')));
        stream.on('error', () => resolve(null));
    });
}

async function scanDirectory(dirPath, sizeThresholdMB, executeLive = false, logFile = 'js_cleanup_audit.log') {
    const absoluteTarget = path.resolve(dirPath);

    let targetStats;
    try {
        targetStats = fs.statSync(absoluteTarget);
    } catch (err) {
        console.error(`[Error] Target directory '${dirPath}' does not exist.`);
        process.exit(1);
    }
    if (!targetStats.isDirectory()) {
        console.error(`[Error] Target '${dirPath}' is not a directory.`);
        process.exit(1);
    }
    if (fs.existsSync(path.join(absoluteTarget, PROTECT_MARKER))) {
        console.error(`[Error] Target is locked by a ${PROTECT_MARKER} marker.`);
        process.exit(1);
    }

    const sizeLimitBytes = sizeThresholdMB * 1024 * 1024;

    let totalFreedBytes = 0;
    const itemsToRemove = [];
    const sizeGroups = new Map();

    console.log(`=== Node.js Storage Purger Scan: ${absoluteTarget} ===`);
    console.log(`Mode: ${executeLive ? 'LIVE_EXECUTION' : 'DRY_RUN (Safe Preview)'}`);

    function walkSync(currentDir, inJunk) {
        let entries = [];
        try {
            entries = fs.readdirSync(currentDir, { withFileTypes: true });
        } catch (err) {
            return;
        }
        for (const entry of entries) {
            const fullPath = path.join(currentDir, entry.name);
            if (entry.isDirectory()) {
                if (!isSkippedDir(fullPath, entry.name)) {
                    walkSync(fullPath, inJunk || CODING_JUNK_DIRS.has(entry.name));
                }
            } else if (entry.isFile()) {
                try {
                    const stats = fs.statSync(fullPath);
                    const ext = path.extname(entry.name).toLowerCase();
                    // 0. 編程垃圾——快取目錄內容與 .pyc/.pyo 編譯殘留
                    if (inJunk || CODING_JUNK_EXTENSIONS.has(ext)) {
                        itemsToRemove.push({ path: fullPath, size: stats.size, reason: 'Coding Junk' });
                        totalFreedBytes += stats.size;
                        continue;
                    }
                    // 1. 暫存檔
                    if (TEMP_EXTENSIONS.has(ext)) {
                        itemsToRemove.push({ path: fullPath, size: stats.size, reason: 'Temp File' });
                        totalFreedBytes += stats.size;
                        continue;
                    }
                    // 2. 超過限制的大檔案
                    if (stats.size > sizeLimitBytes) {
                        itemsToRemove.push({ path: fullPath, size: stats.size, reason: `Over Size Limit (${sizeThresholdMB}MB)` });
                        totalFreedBytes += stats.size;
                        continue;
                    }
                    // 3. 收集「同副檔名 + 同容量」候選待驗證 Hash——複合鍵讓群組
                    //    更小,需要 Hash 的檔案更少(省記憶體與 I/O)
                    if (stats.size >= DUP_MIN_BYTES) {
                        const key = ext + '|' + stats.size;
                        if (!sizeGroups.has(key)) {
                            sizeGroups.set(key, { size: stats.size, paths: [] });
                        }
                        sizeGroups.get(key).paths.push(fullPath);
                    }
                } catch (err) {
                    // 權限缺失或檔名例外跳過
                }
            }
        }
    }

    walkSync(absoluteTarget, false);

    // 比對重複檔案 (2-Pass: 僅對「同副檔名 + 同容量」群組計算 Hash)
    for (const group of sizeGroups.values()) {
        const { size, paths: filePaths } = group;
        if (filePaths.length < 2) continue;
        const hashMap = new Map();
        for (const filePath of filePaths) {
            const hash = await getFileHashAsync(filePath);
            if (!hash) continue;
            if (hashMap.has(hash)) {
                itemsToRemove.push({ path: filePath, size, reason: `Duplicate of ${path.basename(hashMap.get(hash))}` });
                totalFreedBytes += size;
            } else {
                hashMap.set(hash, filePath);
            }
        }
    }

    // 輸出 Log 報告
    const logLines = [
        '=== VERITAS JS PURGER AUDIT LOG ===',
        `Timestamp: ${new Date().toISOString()}`,
        `Target: ${absoluteTarget}`,
        `Execution Mode: ${executeLive ? 'LIVE_DELETE' : 'DRY-RUN'}`,
        'Action Items:'
    ];

    for (const item of itemsToRemove) {
        const line = `[${item.reason}] ${formatSize(item.size)} -> ${item.path}`;
        logLines.push(line);
        console.log(line);
        if (executeLive) {
            try {
                fs.unlinkSync(item.path);
            } catch (err) {
                logLines.push(`   └─ Purge Failed: ${err.message}`);
            }
        }
    }

    logLines.push(`\nSummary: Total ${itemsToRemove.length} files. Space freed: ${formatSize(totalFreedBytes)}`);

    fs.writeFileSync(logFile, logLines.join('\n'), 'utf-8');
    console.log('Scan Complete.');
    console.log(`Target Purge Count: ${itemsToRemove.length} files`);
    console.log(`Potential Space Released: ${formatSize(totalFreedBytes)}`);
    console.log(`Audit saved to ${path.resolve(logFile)}`);
}

// 讀取命令列參數執行
function parseArgs(argv) {
    const opts = { dir: null, maxMB: 200, execute: false, log: 'js_cleanup_audit.log' };
    for (let i = 0; i < argv.length; i++) {
        const arg = argv[i];
        if (arg === '--execute') {
            opts.execute = true;
        } else if (arg === '--max-mb') {
            opts.maxMB = parseFloat(argv[++i]);
        } else if (arg.startsWith('--max-mb=')) {
            opts.maxMB = parseFloat(arg.split('=')[1]);
        } else if (arg === '--log') {
            opts.log = argv[++i];
        } else if (arg.startsWith('--log=')) {
            opts.log = arg.split('=')[1];
        } else if (arg === '--dir') {
            opts.dir = argv[++i];
        } else if (!arg.startsWith('--') && opts.dir === null) {
            opts.dir = arg;
        }
    }
    if (opts.dir === null) opts.dir = '.';
    if (!Number.isFinite(opts.maxMB) || opts.maxMB <= 0) {
        console.error('[Error] Invalid --max-mb value.');
        process.exit(1);
    }
    return opts;
}

const opts = parseArgs(process.argv.slice(2));
scanDirectory(opts.dir, opts.maxMB, opts.execute, opts.log);
